File: go/pickler_combinators.go
package combinators

type pu[A any] func(A)

func pair[A, B any](pa pu[A], pb pu[B]) pu[struct{ A; B }] {
	return func(p struct{ A; B }) {
		pa(p.A)
		pb(p.B)
	}
}

func triple[A, B, C any](pa pu[A], pb pu[B], pc pu[C]) pu[struct{ A; B; C }] {
	return func(p struct{ A; B; C }) {
		pa(p.A)
		pb(p.B)
		pc(p.C)
	}
}

func quad[A, B, C, D any](pa pu[A], pb pu[B], pc pu[C], pd pu[D]) pu[struct{ A; B; C; D }] {
	return func(p struct{ A; B; C; D }) {
		pa(p.A)
		pb(p.B)
		pc(p.C)
		pd(p.D)
	}
}

func wrap[A, B any](i func(A) B, j func(B) A) pu[B] {
	return func(b B) {
		pu[A](j(b))(i(j(b)))
	}
}

func zeroTo(n int) pu[int] {
	if n == 0 {
		return func(int) {}
	}
	return wrap(func(i int) int { return i/256 }, func(i int) int { return i%256 })(zeroTo(n / 256))
}

func unit[A any]() pu[A] {
	return func(A) {}
}

var char = wrap(func(c byte) rune { return rune(c) }, func(r rune) byte { return byte(r) })(zeroTo(255))
var bool = wrap(func(b bool) int { if b { return 1 } else { return 0 } }, func(i int) bool { return i != 0 })(zeroTo(1))

func nat() pu[int] {
	half := 128
	return func(n int) {
		if n < half {
			pu[int](func(int) {})(n)
		} else {
			wrap(func(i int) int { return half + i%half }, func(i int) int { return (i - half) / half })(nat())(n)
		}
	}
}

func fixedList[A any](pa pu[A], n int) pu[[]A] {
	if n == 0 {
		return func([]A) {}
	}
	return wrap(func(p struct{ A; []A }) []A { return append([]A{p.A}, p.[]A...) }, func(l []A) struct{ A; []A } { return struct{ A; []A }{l[0], l[1:]} })(pair(pa, fixedList(pa, n-1)))
}

func list[A any](pa pu[A]) pu[[]A] {
	return func(l []A) {
		n := len(l)
		nat()(n)
		fixedList(pa, n)(l)
	}
}

var string = list(char)

func alt[A any](tag func(A) int, ps []pu[A]) pu[A] {
	if len(ps) == 0 {
		panic("alt: empty list")
	}
	return func(a A) {
		n := tag(a)
		ps[n](a)
	}
}

func pMaybe[A any](pa pu[A]) pu[*A] {
	return alt(func(p *A) int {
		if p == nil {
			return 0
		}
		return 1
	}, []pu[*A]{
		func(*A) {},
		wrap(func(a A) *A { return &a }, func(p *A) A { return *p })(pa),
	})
}

func pEither[A, B any](pa pu[A], pb pu[B]) pu[either] {
	type either struct {
		isLeft bool
		left   A
		right  B
	}
	return alt(func(e either) int {
		if e.isLeft {
			return 0
		}
		return 1
	}, []pu[either]{
		wrap(func(a A) either { return either{true, a, zero(B)} }, func(e either) A { return e.left })(pa),
		wrap(func(b B) either { return either{false, zero(A), b} }, func(e either) B { return e.right })(pb),
	})
}

File: rust/src/pickler_combinators.rs
type Pu<A> = Box<dyn FnOnce(A)>;

pub fn pair<A, B>(pa: Pu<A>, pb: Pu<B>) -> Pu<(A, B)> {
    Box::new(move |(a, b)| {
        pa(a);
        pb(b);
    })
}

pub fn triple<A, B, C>(pa: Pu<A>, pb: Pu<B>, pc: Pu<C>) -> Pu<(A, B, C)> {
    Box::new(move |(a, b, c)| {
        pa(a);
        pb(b);
        pc(c);
    })
}

pub fn quad<A, B, C, D>(pa: Pu<A>, pb: Pu<B>, pc: Pu<C>, pd: Pu<D>) -> Pu<(A, B, C, D)> {
    Box::new(move |(a, b, c, d)| {
        pa(a);
        pb(b);
        pc(c);
        pd(d);
    })
}

pub fn wrap<A, B>(i: impl Fn(A) -> B + 'static, j: impl Fn(B) -> A + 'static) -> Pu<B>
where
    A: 'static,
    B: 'static,
{
    Box::new(move |b| {
        let a = j(b);
        let b = i(a);
    })
}

pub fn zero_to(n: u32) -> Pu<u32> {
    if n == 0 {
        Box::new(|_| {})
    } else {
        wrap(
            |i| i / 256,
            |i| i % 256,
        )(zero_to(n / 256))
    }
}

pub fn unit<A>() -> Pu<A> {
    Box::new(|_| {})
}

pub fn char() -> Pu<char> {
    wrap(
        |c| c as u32 as char,
        |r| r as u32,
    )(zero_to(255))
}

pub fn bool() -> Pu<bool> {
    wrap(
        |b| if b { 1 } else { 0 },
        |i| i != 0,
    )(zero_to(1))
}

pub fn nat() -> Pu<u32> {
    let half = 128;
    Box::new(move |n| {
        if n < half {
            unit()(n);
        } else {
            wrap(
                |i| half + i % half,
                |i| (i - half) / half,
            )(nat())(n);
        }
    })
}

pub fn fixed_list<A>(pa: Pu<A>, n: usize) -> Pu<Vec<A>> {
    if n == 0 {
        Box::new(|_| Vec::new())
    } else {
        wrap(
            |p: (A, Vec<A>)| {
                let mut v = p.1;
                v.insert(0, p.0);
                v
            },
            |mut v: Vec<A>| (v.remove(0), v),
        )(pair(pa, fixed_list(pa, n - 1)))
    }
}

pub fn list<A>(pa: Pu<A>) -> Pu<Vec<A>> {
    Box::new(move |v| {
        let n = v.len();
        nat()(n as u32);
        fixed_list(pa, n)(v);
    })
}

pub fn string() -> Pu<String> {
    list(char())
}

pub fn alt<A>(tag: impl Fn(&A) -> usize + 'static, ps: Vec<Pu<A>>) -> Pu<A> {
    if ps.is_empty() {
        panic!("alt: empty list");
    }
    Box::new(move |a| {
        let n = tag(&a);
        ps[n](a);
    })
}

pub fn p_maybe<A>(pa: Pu<A>) -> Pu<Option<A>> {
    alt(
        |p: &Option<A>| p.is_none() as usize,
        vec![
            Box::new(|_| None),
            wrap(
                |a| Some(a),
                |o| o.unwrap(),
            )(pa),
        ],
    )
}

pub fn p_either<A, B>(pa: Pu<A>, pb: Pu<B>) -> Pu<Either<A, B>> {
    alt(
        |e: &Either<A, B>| e.is_left() as usize,
        vec![
            wrap(
                |a| Either::Left(a),
                |e| match e {
                    Either::Left(a) => a,
                    Either::Right(_) => panic!("impossible"),
                },
            )(pa),
            wrap(
                |b| Either::Right(b),
                |e| match e {
                    Either::Left(_) => panic!("impossible"),
                    Either::Right(b) => b,
                },
            )(pb),
        ],
    )
}